Generation/src/components/HepMCFileReaderTool.cpp

```diff
@@ -40,13 +40,20 @@ StatusCode HepMCFileReader::open(const std::string& filename) {
   return StatusCode::SUCCESS;
 }
 
-HepMC::GenEvent* HepMCFileReader::readNextEvent() {
-  Assert(nullptr != m_file, "Invalid input file!");
-  // this function is to be used in conjunction with
-  // the data service or some other memory manager
-  HepMC::GenEvent* tmpEvent = new HepMC::GenEvent();
-  m_file->fill_next_event(tmpEvent);
-  return tmpEvent;
+StatusCode HepMCFileReader::readNextEvent(HepMC::GenEvent& event) {
+  if(!m_file->fill_next_event(&event)) {
+    if (m_file->rdstate() == std::ios::eofbit) {
+        error() << "Error reading HepMC file" << endmsg;
+        return StatusCode::FAILURE;
+    }
+    // Reached the end of the file: Stop the run
+    IEventProcessor* eventProcessor;
+    service("ApplicationMgr",eventProcessor);
+    eventProcessor->stopRun();
+    error() << "Premature end of file: Please set the number of events according to hepMC file." << endmsg;
+    return Error("Reached end of file before finished processing");
```

Comment:
Here, we could not return the error but only do the `eventProcessor->stopRun()` to stop after the current event. However, I didn't like that because it could mean that we keep the last event where some of the pile-up events can't be read from the file.